server/src/test/java/com/microsoft/java/bs/core/internal/server/BuildTargetServerIntegrationTest.java

```diff
@@ -53,79 +176,157 @@ static void beforeClass() {
+    return matching.get(0);
+  }
+
+  @Test
+  void testCompilingSingleProjectServer() {
+    withNewTestServer("junit5-jupiter-starter-gradle", (gradleBuildServer, client) -> {
+      // get targets
+      WorkspaceBuildTargetsResult buildTargetsResult = gradleBuildServer.workspaceBuildTargets()
+          .join();
+      List<BuildTargetIdentifier> btIds = buildTargetsResult.getTargets().stream()
+          .map(BuildTarget::getId)
+          .collect(Collectors.toList());
+      assertEquals(2, btIds.size());
+
+      // check dependency sources
+      DependencySourcesParams dependencySourcesParams = new DependencySourcesParams(btIds);
+      DependencySourcesResult dependencySourcesResult = gradleBuildServer
+          .buildTargetDependencySources(dependencySourcesParams).join();
+      assertEquals(2, dependencySourcesResult.getItems().size());
+      List<String> allSources = dependencySourcesResult.getItems().stream()
+          .flatMap(item -> item.getSources().stream()).collect(Collectors.toList());
+      assertTrue(allSources.stream().anyMatch(source -> source.endsWith("-sources.jar")));
+
+      // check dependency modules
+      DependencyModulesParams dependencyModulesParams = new DependencyModulesParams(btIds);
+      DependencyModulesResult dependencyModulesResult = gradleBuildServer
+              .buildTargetDependencyModules(dependencyModulesParams).join();
+      assertEquals(2, dependencyModulesResult.getItems().size());
+      List<MavenDependencyModuleArtifact> allArtifacts = dependencyModulesResult.getItems().stream()
+              .flatMap(item -> item.getModules().stream())
+              .filter(dependencyModule -> "maven".equals(dependencyModule.getDataKind()))
+              .map(dependencyModule -> JsonUtils.toModel(dependencyModule.getData(),
+                  MavenDependencyModule.class))
+              .flatMap(mavenDependencyModule -> mavenDependencyModule.getArtifacts().stream())
+              .filter(artifact -> "sources".equals(artifact.getClassifier()))
+              .collect(Collectors.toList());
+      assertTrue(allArtifacts.stream().anyMatch(artifact ->
+          artifact.getUri().endsWith("-sources.jar")));
+
+      // clean targets
+      CleanCacheParams cleanCacheParams = new CleanCacheParams(btIds);
+      CleanCacheResult cleanResult = gradleBuildServer
+          .buildTargetCleanCache(cleanCacheParams).join();
+      assertTrue(cleanResult.getCleaned());
+      client.waitOnMessages(2, 2, 1);
+      client.clearMessages();
+
+      // compile targets
+      CompileParams compileParams = new CompileParams(btIds);
+      CompileResult compileResult = gradleBuildServer.buildTargetCompile(compileParams).join();
+      assertEquals(StatusCode.OK, compileResult.getStatusCode());
+      client.waitOnMessages(2, 2, 1);
+      BuildTargetIdentifier testBt = getBt(btIds, "/junit5-jupiter-starter-gradle/", "test");
```

Comment:
@Arthurm1 Does this case pass on your side? Due to the current implementation at: https://github.com/microsoft/build-server-for-gradle/blob/59b80ec5d8d29174ada027449ab330483e656b8f/server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java#L121-L126, it should fail.

It fails on my machine, but I have no idea why it passes in GitHub Actions.

Reply:
@jdneo Yes it passes on mine.  What line does it fail on - 324?  The messages I have at that point are...
![image](https://github.com/microsoft/build-server-for-gradle/assets/5612295/699f6ff5-b6b8-4540-a8a0-6f8fcd2ba58a)
It's possible you have additional messages.  I noticed that the `CompletableFuture.runAsync(this::reloadWorkspace);` triggered after a compile will cause more messages to be sent to the client but that's normally too slow to register.  I have fix for this and can submit a PR, but it's only ever happened to me when I've sat on a breakpoint for a while.

Reply:
I suspect this limitation causes the failure.

Target id for both `main` and `test` are sent to be compiled. But because of https://github.com/microsoft/build-server-for-gradle/blob/59b80ec5d8d29174ada027449ab330483e656b8f/server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java#L126

only `main` target id is set to the compile reporter, which fails line 326?

Reply:
@jdneo This makes sense - it could be failing on your machine due to message ordering differences.  I've finished off #116 which should fix this if you'd like to review it.  It reports on progress-per-target rather than just reporting everything on the first target.

Reply:
Cool! We can eliminate a protocol violation then. I'll review it these days. Thank you!